src/dos/cdrom_image.cpp

```diff
@@ -1174,6 +1174,235 @@ static std::string dirname(char* file)
+			return false;
+		}
+		track.start = *start_sector;
+		const auto start_offset = read_native_u64_at(file, current_track_offset + 40);
+		if (!start_offset) {
+			LOG_ERR("CDROM: Invalid MDS file");
+			return false;
+		}
+		const auto num_files = read_native_u32_at(file, current_track_offset + 48);
+		if (!num_files) {
+			LOG_ERR("CDROM: Invalid MDS file");
+			return false;
+		}
+		track.skip = check_cast<uint32_t>(*start_offset);
+		if (*num_files != 1) {
+			// According to comments in cdemu, DVD tracks can be split into multiple files.
+			// We don't care about DVDs and CDROM should always be 1 file per track.
+			LOG_ERR("CDROM: %u files in track %hhu. Must be exactly 1.", *num_files, track.number);
+			return false;
+		}
+		const auto footer_offset = read_native_u32_at(file, current_track_offset + 52);
+		if (!footer_offset || *footer_offset == 0) {
+			LOG_ERR("CDROM: Invalid MDS file");
+			return false;
+		}
+		const auto length = read_native_u32_at(file, *extra_offset + 4);
+		if (!length) {
+			LOG_ERR("CDROM: Invalid MDS file");
+			return false;
+		}
+		track.length = *length;
+		const auto filename_offset = read_native_u32_at(file, *footer_offset);
+		if (!filename_offset || *filename_offset == 0) {
+			LOG_ERR("CDROM: Invalid MDS file");
+			return false;
+		}
+		file.seekg(*filename_offset);
+		if (file.fail()) {
+			LOG_ERR("CDROM: Invalid MDS file");
+			return false;
+		}
+		uint32_t prev_track = tracks.empty() ? 0 : tracks.back().number; //-V807
+		uint32_t prev_sector = tracks.empty() ? 0 : tracks.back().start + tracks.back().length;
+		if (track.number != prev_track + 1 || track.start < prev_sector) {
+			LOG_ERR("CDROM: Non-contigious track found in MDS file");
+			return false;
+		}
+		std::u16string utf16_string = {};
+		while (true) {
+			uint16_t wide_char = 0;
+			file.read(reinterpret_cast<char*>(&wide_char), sizeof(wide_char));
+			if (wide_char == 0) {
+				break;
+			}
+			utf16_string.push_back(le16_to_host(wide_char));
+		}
+		std_fs::path mdf_filename(utf16_string);
+		if (mdf_filename.empty()) {
+			LOG_ERR("CDROM: Missing MDF filename");
+			return false;
+		}
+		if (mdf_filename == "*.mdf") {
+			mdf_filename = mds_path;
+			mdf_filename.replace_extension("mdf");
+		} else {
+			mdf_filename = mds_path.parent_path() / mdf_filename;
+		}
+		if (const auto iter = track_map.find(mdf_filename); iter != track_map.end()) {
+			track.file = iter->second;
+		} else {
+			bool error = false;
+			track.file = std::make_shared<BinaryFile>(mdf_filename, error);
+			if (error) { //-V547
+				LOG_ERR("CDROM: Failed to open MDF file: %s", mdf_filename.string().c_str());
+				return false;
+			}
+			track_map.emplace(mdf_filename, track.file);
+		}
+		tracks.push_back(track);
+	}
+	if (tracks.empty()) {
+		LOG_ERR("CDROM: Failed to find any tracks");
+		return false;
+	}
+	Track lead_out = {};
+	lead_out.start = tracks.back().start + tracks.back().length;
+	tracks.push_back(lead_out);
+	return true;
+}
```

Comment:
This method is way too big. Can you break it up into 2-3 or more smaller methods? A trivial candidate is the stuff happening in the for loop, but even that is too big.